### ISSUE TITLE
WASM Binary Fails to Create Directory

### ISSUE THREAD

**Issue by Noordfrees** [FreeRCT/FreeRCT#314]
The online version of FreeRCT fails to start with the error:
```
Failed creating directory 'wasm_userdata/save/' [freerct.js:1:38688](https://freerct.net/play/freerct.js)
Aborted(native code called abort())
```

To reproduce: Go to https://freerct.net/play/play?guest=true and see the browser's javascript console. (edit: I changed the website to serve an older, working commit until this is resolved.)

This seems to be a regression of #305 @rsf92

**Comment by rsf92** [comment]
Hmm, ok I'll check as soon as I can at home. Is there a way that one can check the issue now that the server has been rollbacked?

**Comment by Noordfrees** [comment]
You can compile the WASM binary locally. You'll need to [install the Emscripten toolkit](https://emscripten.org/docs/getting_started/downloads.html), then compile FreeRCT using
```sh
mkdir wasm  # Use a clean build directory
cd wasm
source /path/to/emsdk/emsdk_env.sh  # Source the file appropriate to your shell
cmake -DWEBASSEMBLY=ON ..
make -j1  # Use only 1 core for WASM builds
```
This will give you a file `wasm/bin/freerct.html` which you can open and run in your browser.

**Comment by rsf92** [comment]
Fine, I have the html file, but I don't really know how to debug this... any hints? Using the browser's code inspection does not seem to be enough

**Comment by rsf92** [comment]
I get at different error when I load it, although I'm pretty sure the cause for this issue is that I used `std::filesystem::create_directory` and for this case it needs to create both directories, so `std::filesystem::creater_directories` should be used. Should I just test it with the regular game or do you have any other solution in mind?

**Comment by Noordfrees** [comment]
If you provide a branch, I can compile and deploy it on the website so we can test it there live

**Comment by rsf92** [comment]
Never mind, my mistake, it was already using create_directories, I'll try to sort out the problems I have or come up with some other possible explanation

**Comment by rsf92** [comment]
Just to clarify, wasm_userdata is supposed to be a path relative to the html file, right? Would adding a ./ break anything in principle?

**Comment by Noordfrees** [comment]
No, WASM's "filesystem" doesn't actually correspond to anything on the physical disk at all. WASM binaries are executed by the browser's JavaScript engine which has no access to the computer's file system for security reasons. All files used by WASM exist purely in your computer's RAM. This applies both to static data like the RCD files that is downloaded at startup and to userdata such as savegames. Where the virtual working directory and the virtual file storage actually are, I have no idea. But if you need any WASM-specific logging output, just make a branch and we'll put it up, taking the website's Play section down for troubleshooting for a few days wouldn't be that much of a problem IMO.
